sensor/common/clusterentities/store_endpoints.go

```diff
@@ -0,0 +1,254 @@
+	reverseHistoricalEndpoints map[string]map[net.NumericEndpoint]*entityStatus
+}
+
+func newEndpointsStoreWithMemory(numTicks uint16) *endpointsStore {
+	store := &endpointsStore{entitiesMemorySize: numTicks}
+	store.mutex.Lock()
+	defer store.mutex.Unlock()
+	store.initMapsNoLock()
+	return store
+}
+
+func (e *endpointsStore) initMapsNoLock() {
+	e.endpointMap = make(map[net.NumericEndpoint]map[string]map[EndpointTargetInfo]struct{})
+	e.reverseEndpointMap = make(map[string]map[net.NumericEndpoint]struct{})
+
+	e.reverseHistoricalEndpoints = make(map[string]map[net.NumericEndpoint]*entityStatus)
+	e.historicalEndpoints = make(map[net.NumericEndpoint]map[string]map[EndpointTargetInfo]*entityStatus)
+}
+
+func (e *endpointsStore) resetMaps() {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	// Maps holding historical data must not be wiped on reset! Instead, all entities must be marked as historical.
+	// Must be called before the respective source maps are wiped!
+	// Performance optimization: no need to handle history if history is disabled
+	if e.entitiesMemorySize == 0 {
+		e.initMapsNoLock()
+		return
+	}
+	e.markAllEndpointsHistorical()
+
+	e.endpointMap = make(map[net.NumericEndpoint]map[string]map[EndpointTargetInfo]struct{})
+	e.reverseEndpointMap = make(map[string]map[net.NumericEndpoint]struct{})
+	e.updateMetricsNoLock()
+}
+
+func (e *endpointsStore) updateMetricsNoLock() {
+}
+
+func (e *endpointsStore) RecordTick() {
+	for deploymentID, m := range e.reverseHistoricalEndpoints {
+		for endpoint, status := range m {
+			status.recordTick()
+			// Remove all historical entries that expired in this tick.
+			e.removeHistoricalExpiredDeploymentEndpoints(deploymentID, endpoint)
+		}
+	}
+}
+
+func (e *endpointsStore) Apply(
+	updates map[string]*EntityData,
+	incremental bool,
+	incPublicIPRef func(addr net.IPAddress),
+	decPublicIPRef func(addr net.IPAddress),
+) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	e.applyNoLock(updates, incremental, incPublicIPRef, decPublicIPRef)
+}
+
+func (e *endpointsStore) applyNoLock(
+	updates map[string]*EntityData,
+	incremental bool,
+	incPublicIPRef func(addr net.IPAddress),
+	decPublicIPRef func(addr net.IPAddress),
+) {
+	defer e.updateMetricsNoLock()
+	if !incremental {
+		for deploymentID := range updates {
+			e.purgeNoLock(deploymentID, decPublicIPRef)
+		}
+	}
+	for deploymentID, data := range updates {
+		if data == nil {
+			continue
+		}
+		e.applySingleNoLock(deploymentID, *data, incPublicIPRef)
+	}
+}
+
+func (e *endpointsStore) purgeNoLock(deploymentID string, decPublicIPRef func(addr net.IPAddress)) {
+	for ep := range e.reverseEndpointMap[deploymentID] {
+		e.markEndpointHistorical(deploymentID, ep)
+		// For entitiesMemorySize > 0, the deletion of historical expired entries happens after a tick.
+		// If memory is disabled, we should delete historical expired entries immediately.
+		e.removeHistoricalExpiredDeploymentEndpoints(deploymentID, ep)
```

Comment:
I think this is still confusing (potentially buggy). If the memory is 0, `makrEndpointHistorical` will not add anything to the `reverseHistoricalEndpoints`. If the historical maps are empty `removeHistoricalExpiredDeploymentEndpoints` will not call `removeDeploymentEndpoints`. If `removeDeploymentEndpoint` is not called we will never delete the entries in the `endpointMap`.

Maybe I'm missing something but I think this is definitely confusing.

Reply:
I revamped that part, I hope it should be clearer now.